### PR TITLE
8286294 : ForkJoinPool.commonPool().close() spins

### PR DESCRIPTION
**Pull request by DougLea**
Changes ForkJoinPool.close spec and code to trap close as a no-op if called on common pool

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires a CSR request to be approved
- [x] Change must be properly reviewed (1 review required, with at least 1 reviewer)

### Issues
 * [JDK-8286294](https://bugs.openjdk.java.net/browse/JDK-8286294): ForkJoinPool.commonPool().close() spins
 * [JDK-8286341](https://bugs.openjdk.java.net/browse/JDK-8286341): ForkJoinPool.commonPool().close() spins (**CSR**)

### Reviewers
 * [Alan Bateman](https://openjdk.java.net/census#alanb) (@AlanBateman - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/8577/head:pull/8577` \
`$ git checkout pull/8577`

Update a local copy of the PR: \
`$ git checkout pull/8577` \
`$ git pull https://git.openjdk.java.net/jdk pull/8577/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 8577`

View PR using the GUI difftool: \
`$ git pr show -t 8577`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/8577.diff">https://git.openjdk.java.net/jdk/pull/8577.diff</a>

</details>

### CODE REVIEW

**Review by PaulSandoz — COMMENTED** [review]
Changes look good, it will need a CSR.

File: test/jdk/java/util/concurrent/tck/ForkJoinPool19Test.java [hunk 1]
```diff
@@ -492,4 +492,31 @@ public void testAdaptInterruptible_Callable_toString() {
         }
     }
 
+    /**
+     * Implictly closing a new pool using try-with-resources terminates it
```
Comment (Martin-Buchholz) [inline test/jdk/java/util/concurrent/tck/ForkJoinPool19Test.java:496]: Typo "Implictly" - twice

Comment (Martin-Buchholz) [reply test/jdk/java/util/concurrent/tck/ForkJoinPool19Test.java]: Tests in this file are not being executed.  I think you need:

--- a/test/jdk/java/util/concurrent/tck/ForkJoinPool19Test.java
+++ b/test/jdk/java/util/concurrent/tck/ForkJoinPool19Test.java
@@ -55,7 +55,7 @@ public class ForkJoinPool19Test extends JSR166TestCase {
     }
 
     public static Test suite() {
-        return new TestSuite(ForkJoinPool8Test.class);
+        return new TestSuite(ForkJoinPool19Test.class);
     }
 
     /**

Comment (Martin-Buchholz) [reply test/jdk/java/util/concurrent/tck/ForkJoinPool19Test.java]:
```
--- a/test/jdk/java/util/concurrent/tck/ForkJoinPool19Test.java
+++ b/test/jdk/java/util/concurrent/tck/ForkJoinPool19Test.java
@@ -55,7 +55,7 @@ public class ForkJoinPool19Test extends JSR166TestCase {
     }
 
     public static Test suite() {
-        return new TestSuite(ForkJoinPool8Test.class);
+        return new TestSuite(ForkJoinPool19Test.class);
     }
 
     /**
```

Comment (Martin-Buchholz) [reply test/jdk/java/util/concurrent/tck/ForkJoinPool19Test.java]: testLazySubmit will cause jtreg to start hanging.
If you wait patiently for 1000 seconds, you'll get a stack dump.

Comment (DougLea) [reply test/jdk/java/util/concurrent/tck/ForkJoinPool19Test.java]: Thanks; now fixed (and enabled). (The test didn't do what doc said about joined "by a worker")

**Comment by Martin-Buchholz** [comment]
testAdaptInterruptible_Callable_toString belongs in ForkJoinTaskTest.java.  Oh wait, it's already there, commented out!  Why not fix it there?

**Comment by Martin-Buchholz** [comment]
Here's a suggested strengthening of testCloseCommonPool:

- close should have "no effect", not just "not terminate".
- the submitted task will be run by the pool ... eventually  (which suggests that closing the common pool maybe should quiesce the pool before returning)
-  I might merge this with testCommonPoolShutDown
```
    /**
     * Implicitly closing common pool using try-with-resources has no effect.
     */
    public void testCloseCommonPool() {
        ForkJoinTask f = new FibAction(8);
        ForkJoinPool pool;
        try (ForkJoinPool p = pool = ForkJoinPool.commonPool()) {
            p.execute(f);
        }

        assertFalse(pool.isShutdown());
        assertFalse(pool.isTerminating());
        assertFalse(pool.isTerminated());

        String prop = System.getProperty(
            "java.util.concurrent.ForkJoinPool.common.parallelism");
        if (! "0".equals(prop)) {
            f.join();
            checkCompletedNormally(f);
        }
    }
```

File: src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java [hunk 2]
```diff
@@ -3529,6 +3529,54 @@ public boolean awaitQuiescence(long timeout, TimeUnit unit) {
         return (helpQuiescePool(this, unit.toNanos(timeout), false) > 0);
     }
 
+    /**
+     * Unless this is the {@link #commonPool()}, initiates an orderly
```
Comment (Martin-Buchholz) [inline src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java:3533]: clearer everywhere is
{@linkplain #commonPool common pool}

File: src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java [hunk 3]
```diff
@@ -3529,6 +3529,54 @@ public boolean awaitQuiescence(long timeout, TimeUnit unit) {
...
+     * shutdown in which previously submitted tasks are executed, but
+     * no new tasks will be accepted, and waits until all tasks have
+     * completed execution and the executor has terminated.
```
Comment (Martin-Buchholz) [inline src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java:3536]: slightly clearer is
s/executor/pool/

File: src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java [hunk 4]
```diff
@@ -3529,6 +3529,54 @@ public boolean awaitQuiescence(long timeout, TimeUnit unit) {
...
+     * #commonPool()}, this method has no effect on execution, and
+     * does not wait. Otherwise, if interrupted while waiting, this
+     * method stops all executing tasks as if by invoking {@link
+     * #shutdownNow()}. It then continues to wait until all actively
+     * executing tasks have completed. Tasks that were awaiting
+     * execution are not executed. The interrupt status will be
+     * re-asserted before this method returns.
+     *
+     * @throws SecurityException if a security manager exists and
+     *         shutting down this ExecutorService may manipulate
```
Comment (Martin-Buchholz) [inline src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java:3548]: clearer and more consistent is
s/this ExecutorService/this pool/

File: src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java [hunk 5]
```diff
@@ -3529,6 +3529,54 @@ public boolean awaitQuiescence(long timeout, TimeUnit unit) {
...
+    @Override
+    public void close() {
+        if ((config & ISCOMMON) == 0) {
+            boolean terminated = tryTerminate(false, false);
+            if (!terminated) {
+                shutdown();
+                boolean interrupted = false;
+                while (!terminated) {
+                    try {
+                        terminated = awaitTermination(1L, TimeUnit.DAYS);
```
Comment (Martin-Buchholz) [inline src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java:3565]: I would use untimed wait

**Comment by DougLea** [comment]
@Martin-Buchholz thanks for test improvements. I'm leaving further FJP tweaks for some other time to avoid possibility of problems after already merging with loom commit

**Review by AlanBateman — APPROVED** [review]

**Pull request closed without merging**